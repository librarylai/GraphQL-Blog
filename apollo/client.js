import { useMemo } from 'react'
import { split, ApolloClient, InMemoryCache } from '@apollo/client'
import { createClient } from 'graphql-ws'
import { getMainDefinition } from '@apollo/client/utilities'


let apolloClient

function createIsomorphLink() {
  // server / statie 端 (SSR,SSG)
  if (typeof window === 'undefined') {
    const { SchemaLink } = require('@apollo/client/link/schema')
    const { schema } = require('./schema')
    const { getDB } = require('../mongodb')
    return new SchemaLink({
      schema,
      context: async () => {
        let db = await getDB()
        return {
          blogDB: db.collection('blog'),
          userDB: db.collection('user'),
        }
      },
    })
  } else {
    // Client 端 (CSR)
    const { HttpLink } = require('@apollo/client/link/http')
    const { GraphQLWsLink } = require('@apollo/client/link/subscriptions')
    const wsLink = new GraphQLWsLink(
      createClient({
        url: '/api/graphql',
      })
    )
    const httpLink = new HttpLink({
      uri: '/api/graphql',
      credentials: 'same-origin',
    })
    const splitLink = split(
      ({ query }) => {
        const definition = getMainDefinition(query)
        // 如果 operation 為 subscription 則 return true
        return definition.kind === 'OperationDefinition' && definition.operation === 'subscription'
      },
      wsLink, // 如果是 true 則為 wsLink
      httpLink // 如果是 false 則為 httpLink
    )
    return splitLink
  }
}

function createApolloClient() {
  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: createIsomorphLink(),
    cache: new InMemoryCache(),
  })
}

export function initializeApollo(initialState = null) {
  const _apolloClient = apolloClient ?? createApolloClient()

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    _apolloClient.cache.restore(initialState)
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === 'undefined') return _apolloClient
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient

  return _apolloClient
}

export function useApollo(initialState) {
  const store = useMemo(() => initializeApollo(initialState), [initialState])
  return store
}
