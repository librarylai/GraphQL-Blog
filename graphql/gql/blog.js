import gql from 'graphql-tag'

export const ALL_POST_QUERY = gql`
  query AllPostQuery {
    viewAllPost {
      id
      title
      content
    }
  }
`
export const ADD_POST_QUERY = gql`
  mutation AddPost($title: String, $content: String) {
    addPost(title: $title, content: $content) {
      id
      title
    }
  }
`
export const UPDATE_POST_QUERY = gql`
  mutation UpdatePost($postId: ID, $title: String, $content: String) {
    updatePost(postId: $postId, title: $title, content: $content){
      id
      title
    }
  }
`
export const DELETE_POST_QUERY = gql`
  mutation DeletePost($postId: ID) {
    deletePost(postId: $postId) {
      id
      title
    }
  }
`