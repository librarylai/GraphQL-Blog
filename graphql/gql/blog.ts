import gql from 'graphql-tag'

export const ALL_POST_QUERY = gql`
  query AllPostQuery {
    viewAllPost {
      id
      title
      content
      author {
        id
        name
      }
    }
  }
`
export const ADD_POST_QUERY = gql`
  mutation AddPost($title: String, $content: String, $authorId: ID) {
    addPost(title: $title, content: $content, authorId: $authorId) {
      id
      title
      content
      author {
        id
        name
      }
    }
  }
`
export const UPDATE_POST_QUERY = gql`
  mutation UpdatePost($postId: ID, $title: String, $content: String, $authorId: ID) {
    updatePost(postId: $postId, title: $title, content: $content, authorId: $authorId) {
      id
      title
      content
      author {
        id
        name
      }
    }
  }
`
export const DELETE_POST_QUERY = gql`
  mutation DeletePost($postId: ID) {
    deletePost(postId: $postId) {
      id
      title
      content
      author {
        id
        name
      }
    }
  }
`
// 新增評論
export const ADD_COMMENT = gql`
  mutation AddComment($postId: ID, $comment: String, $authorId: ID) {
    addComment(postId: $postId, comment: $comment, authorId: $authorId) {
      comment
    }
  }
`
