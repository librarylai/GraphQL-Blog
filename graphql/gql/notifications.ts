import gql from 'graphql-tag'

export const ALL_NOTIFICATIONS_QUERY = gql`
  query fetchNotifications {
    user {
      # 排序過後
      sortNotifications @client {
        id
        content {
          title
          subTitle
          createTime
          isRead
        }
      }
      # 原始資料
      notifications {
        id
        content {
          title
          subTitle
          createTime
          isRead
        }
      }
    }
  }
`

export const ALL_NOTIFICATIONS_TITLE_QUERY = gql`
  query fetchNotificationsTitle {
    user {
      notifications {
        id
        content {
          title
        }
      }
    }
  }
`

export const ALL_NOTIFICATIONS_SUBTITLE_QUERY = gql`
  query fetchNotificationsSubTitle {
    user {
      notifications {
        id
        content {
          subTitle
        }
      }
    }
  }
`

export const UPDATE_NOTIFICATION = gql`
  mutation updateNotification($id: ID) {
    updateNotification(notificationId: $id) {
      status
    }
  }
`
