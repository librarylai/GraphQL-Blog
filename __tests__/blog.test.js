import { prettyDOM, render, waitFor, screen } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import Index from '@/pages/index'
import { ALL_POST_QUERY, ADD_POST_QUERY, DELETE_POST_QUERY } from '@/graphql/gql/blog'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'

function createPostData(id, title, content, authorName) {
  return {
    id,
    title,
    content,
    author: {
      id,
      name: authorName,
    },
    __typename: 'Post',
  }
}

// 用來 mock notification request and response
const mockBlogData = [
  {
    request: {
      query: ALL_POST_QUERY,
      // variables : 如果輸入參數為 {} 則 result 為以下....
      variables: {},
    },
    result: {
      data: {
        allPost: [
          createPostData('1', '好想出去玩', '天氣真好', 'library'),
          createPostData('2', '好想買遊戲王卡', '青眼白龍勒～～', 'Zack'),
          createPostData('3', '好想打遊戲', '殺～～～', 'Wang'),
        ],
      },
    },
  },
]

const mockBlogErrorData = [
  {
    request: {
      query: ALL_POST_QUERY,
      variables: {}, // 如果有需要帶入查詢參數，則要加入 variables
    },
    error: new Error('系統發生錯誤，請再重試一次'),
  },
]

const mockAddPostData = [
  {
    request: {
      query: ADD_POST_QUERY,
      // variables 可以想像是 我們去預設當輸入參數為 多少時(variables)，則回傳值會是怎樣...
      // 例如： 當輸入參數為 {title:'測試新增文章' ,content:'測試內容' ,authorId:1} 的話，則回傳 createPostData('1', '測試新增文章', '測試測試', 'library')
      // 因此我們可以設計多個 variables 對應 result 的情況。
      variables: {
        title: '測試新增文章',
        content: '測試內容',
        authorId: 1,
      },
    },
    result: {
      data: {
        addPost: [createPostData('1', '測試新增文章', '測試內容', 'library')],
      },
    },
  },
]

const mockDeletePostData = [
  {
    request: {
      query: DELETE_POST_QUERY,
    },
    result: jest.fn(),
  },
]

// 測試 Loading 情況
it('1. Test Blog Loading Situation', async () => {
  const { container } = render(
    <MockedProvider mocks={mockBlogData}>
      <Index />
    </MockedProvider>
  )
  expect(container).toMatchSnapshot()
})

// 測試 Error 情況
it('2. Test Blog Error Situation', async () => {
  const { container } = render(
    <MockedProvider mocks={mockBlogErrorData}>
      <Index />
    </MockedProvider>
  )
  // 因為我們在操作 GraphQL 是屬於非同步的狀態(Promise-based)，因此需要做一個 Promise 去等待操作完成
  // 這邊有額外使用 react-testing-library 的 waitFor 方法，讓整個測試必須整個 callback 完成後再進行下去(專門處理複雜情況，例如：等待資料回來後再額外進行其他判斷)。
  // 但這邊情境比較單存，因此可以直接用 await new Promise((resolve) => setTimeout(resolve, 0))
  await waitFor(() => new Promise((resolve) => setTimeout(resolve, 0)))
  expect(container).toMatchSnapshot()
})

// 測試資料顯示情況
it('3. Test Blog Data Situation', async () => {
  const { container } = render(
    <MockedProvider mocks={mockBlogData}>
      <Index />
    </MockedProvider>
  )
  // 因為我們在操作 GraphQL 是屬於非同步的狀態(Promise-based)，因此需要做一個 Promise 去等待操作完成
  // 這邊有額外使用 react-testing-library 的 waitFor 方法，讓整個測試必須整個 callback 完成後再進行下去(專門處理複雜情況，例如：等待資料回來後再額外進行其他判斷)。
  // 但這邊情境比較單存，因此可以直接用 await new Promise((resolve) => setTimeout(resolve, 0))
  await waitFor(() => new Promise((resolve) => setTimeout(resolve, 0)))
  expect(container).toMatchSnapshot()
})

// 測試新增 mutation
it('4. Test Add new Blog Situation', async () => {
  const user = userEvent.setup()
  const { container } = render(
    <MockedProvider mocks={mockAddPostData} addTypename={false}>
      <Index />
    </MockedProvider>
  )
  // 找到 新增文章 按鈕
  const addPostButton = await screen.queryByText('新增文章')
  /* PS. 檢測看看是否有正確找到元件 
    這邊可以用 console 搭配 prettyDOM 來檢測看看是否有正確找到 Button
    console.log('addPostButton', prettyDOM(addPostButton))
  */
  // 等待 addPostButton 被點擊後，打開 PostDialog 彈窗
  await user.click(addPostButton)
  // 找到 PostDialog 標題 input
  const postDialogTitleInput = await screen.getByLabelText('postDialog-title').querySelector('input')
  // 找到 PostDialog 內容 textbox
  const postDialogContentInput = await screen.getByLabelText('postDialog-content')
  // 找到 PostDialog 彈窗裡的『送出』按鈕
  const addPostSubmitButton = await screen.queryByText('送出')
  // 標題 填入『測試新增文章』
  await user.type(postDialogTitleInput, '測試新增文章')
  // 內容 填入『測試內容』
  await user.type(postDialogContentInput, '測試內容')
  // 等待 送出按鈕 被點擊
  await user.click(addPostSubmitButton)
  // 等待 mockProvider 非同步執行
  await waitFor(() => new Promise((resolve) => setTimeout(resolve, 0)))
  // 執行 Snapshot
  expect(container).toMatchSnapshot()
})

it('5. Test delete Blog Situation', async () => {
  const { container } = render(
    <MockedProvider mocks={mockAddPostData} addTypename={false}>
      <Index />
    </MockedProvider>
  )
})
