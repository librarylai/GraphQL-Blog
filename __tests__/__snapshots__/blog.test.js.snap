// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. Test Blog Loading Situation 1`] = `
<div>
  <div>
    <div
      class="MuiBox-root css-2imjyh"
    >
      <div
        class="pages__LoadingWrapper-sc-1cvwks4-0 ejApMF"
      >
        Loading....
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      id="addPost"
      tabindex="0"
      type="button"
    >
      新增文章
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`2. Test Blog Error Situation 1`] = `
<div>
  <div>
    <div
      class="MuiBox-root css-2imjyh"
    >
      <div
        class="pages__ErrorWrapper-sc-1cvwks4-1 bOvhcR"
      >
        系統發生錯誤，請再重試一次
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      id="addPost"
      tabindex="0"
      type="button"
    >
      新增文章
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`3. Test Blog Data Situation 1`] = `
<div>
  <div>
    <div
      class="MuiBox-root css-2imjyh"
    >
      <div
        class="MuiBox-root css-bgk70e"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-9nrmgl-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              好想出去玩
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              天氣真好
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              作者：library
            </p>
          </div>
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              評論：
            </div>
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorSecondary css-6oqcxu-MuiInputBase-root-MuiInput-root"
              variant="outlined"
            >
              <input
                class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                placeholder="寫上評論"
                type="text"
                value=""
              />
            </div>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root  css-11qr2p8-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              送出評論
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              編輯
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-6ekzlo-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              刪除
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-bgk70e"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-9nrmgl-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              好想買遊戲王卡
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              青眼白龍勒～～
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              作者：Zack
            </p>
          </div>
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              評論：
            </div>
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorSecondary css-6oqcxu-MuiInputBase-root-MuiInput-root"
              variant="outlined"
            >
              <input
                class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                placeholder="寫上評論"
                type="text"
                value=""
              />
            </div>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root  css-11qr2p8-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              送出評論
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              編輯
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-6ekzlo-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              刪除
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-bgk70e"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-9nrmgl-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              好想打遊戲
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              殺～～～
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              作者：Wang
            </p>
          </div>
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              評論：
            </div>
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorSecondary css-6oqcxu-MuiInputBase-root-MuiInput-root"
              variant="outlined"
            >
              <input
                class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                placeholder="寫上評論"
                type="text"
                value=""
              />
            </div>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root  css-11qr2p8-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              送出評論
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              編輯
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-6ekzlo-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              刪除
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      id="addPost"
      tabindex="0"
      type="button"
    >
      新增文章
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`4. Test Add new Blog Situation 1`] = `
<div
  aria-hidden="true"
>
  <div>
    <div
      class="MuiBox-root css-2imjyh"
    >
      <div
        class="MuiBox-root css-bgk70e"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-9nrmgl-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              測試新增文章
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              測試內容
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              作者：library
            </p>
          </div>
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              評論：
            </div>
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorSecondary css-6oqcxu-MuiInputBase-root-MuiInput-root"
              variant="outlined"
            >
              <input
                class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                placeholder="寫上評論"
                type="text"
                value=""
              />
            </div>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root  css-11qr2p8-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              送出評論
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-1knaqv7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              編輯
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-6ekzlo-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              刪除
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      id="addPost"
      tabindex="0"
      type="button"
    >
      新增文章
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
